Refactor and fix Listen and Dial concurrency logic

FiloSottile opened the pull request:
## Listen

If a connection fully establishes, passing the timeout select in
handleIncoming, but then is never taken out of a full incoming queue,
the goroutine is leaked. handleIncoming will in turn leak waiting for
wg.Done. Canceling the context, closing the listener, or closing the
goprocess don't affect these goroutines. connAcceptBuffer makes this
problem a little less evident, but doesn't fix it.

Also, document the listener background behaviour.

## Dial

If a context cancellation hit after a successful dial but before the send on the done channel, a connection would be leaked. Since almost all operations already supported context, it was cleaner to streamline the operation.

Consider adding context support natively to multiaddr.

FiloSottile committed:
Fix a goroutine leak in accepting connections

If a connection fully establishes, passing the timeout select in
handleIncoming, but then is never taken out of a full incoming queue,
the goroutine is leaked. handleIncoming will in turn leak waiting for
wg.Done. Canceling the context, closing the listener, or closing the
goprocess don't affect these goroutines. connAcceptBuffer makes this
problem a little less evident, but doesn't fix it.

FiloSottile committed:
Streamline Dial logic and fix a connection leak

If a context cancellation hit after a successful dial but before the send on
the done channel, a connection would be leaked. Since almost all operations
already supported context, it was cleaner to streamline the operation.

Consider adding context support natively to multiaddr.

Stebalien commented on dial.go:
```diff
@@ -89,99 +89,77 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
-		logdial["error"] = errOut.Error()
-		logdial["dial"] = "failure"
-		return nil, errOut
+	c2, err := newSecureConn(ctx, d.PrivateKey, c)
```
Hm. Looking into secio, it turns out that it doesn't fully obey this context. We'll have to fix that first. https://github.com/libp2p/go-libp2p-secio/pull/25

Stebalien commented on dial.go:
```diff
@@ -89,99 +89,77 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 		return nil, ipnet.ErrNotInPrivateNetwork
 	}
 
-	var connOut iconn.Conn
```
If we're going to do this this way, `err` should be a named return value. Otherwise, directly returning an error won't set `err` correctly.

FiloSottile replied:
I was using a local variable (which yeah, is not super comfortable) not to make the godoc stutter, but it's a subjective tradeoff. I'll switch.

Stebalien commented on dial.go:
```diff
@@ -89,99 +89,77 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
-		return nil, ctx.Err()
-	case <-done:
-		// whew, finished.
+		err = ctx.Err()
```
With a named return value, you could just return `nil, ctx.Err()` here.

Stebalien commented on dial.go:
```diff
@@ -89,99 +89,77 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 
+	selectResult := make(chan error, 1)
+	go func() {
+		selectResult <- msmux.SelectProtoOrFail(cryptoProtoChoice, maconn)
```
Yeah... this really should take a context.

Stebalien commented on dial.go:
```diff
@@ -89,99 +89,77 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 
-		c2, err := newSecureConn(ctx, d.PrivateKey, c)
+	if d.Protector != nil {
+		maconn, err = d.Protector.Protect(maconn)
```
I am slightly worried that we could try to make a protector that does work here (and needs to either take a context or be called from a go routine. Thoughts @Kubuxu?

Kubuxu replied:
It is possible but protectors currently work lazily, so I am not sure.

Stebalien replied:
Got it. We just need to remember to always do that.

Stebalien commented on listen_test.go:
```diff
@@ -0,0 +1,40 @@
+	"time"
+
+	tu "github.com/libp2p/go-testutil"
+	grc "github.com/whyrusleeping/gorocheck"
```
We'll need to add this to the gx dependency list `gx import github.com/whyrusleeping/gorocheck` (and then answer `n` when it asks you if you'd like to rewrite the paths).

FiloSottile replied:
Done. I'll admit I'm developing without `gx`, as it makes working cross-package a bit awkward, and I noticed CI runs without it anyway. Is that cool?

FiloSottile replied:
(CI pointed out that `gorocheck` was already in `package.json`, so I dropped that commit.)

Stebalien replied:
Oh. Sorry, I thought it was new.

Stebalien reviewed (changes requested):
Awesome! Thanks!

FiloSottile committed:
Expand README and documentation

FiloSottile committed:
Replace WrapFunc with ConnWrapper

The two types are identical and serve the same purpose.

This change is unlikely to break other code. If it does, a type alias
can be used, along with a deprecation notice.

FiloSottile committed:
Use named returns in Dial

FiloSottile commented:
Added some documentation and addressed the review comments on my side.

whyrusleeping commented:
@FiloSottile Thanks! This is great stuff.

> Consider adding context support natively to multiaddr.

What specifically do mean here? We have contexts on multiaddr-net Dials (via `DialContext`). We could look at adding it to the listener, maybe in the `Accept` method?

FiloSottile commented:
Ah, it’s a typo, I meant multistream.

whyrusleeping commented:
Ah, yeah. We should totally add contexts down into multistream. Though the effectiveness is somewhat limited as Read and Write calls don't play nicely.

Stebalien commented:
> Ah, yeah. We should totally add contexts down into multistream. Though the effectiveness is somewhat limited as Read and Write calls don't play nicely.

We can use them in conjunction with goroutines as we do here when negotiating the protocol. Spinning up a goroutine for something like that isn't *that* expensive.

Stebalien commented:
I've fixed the context issue in secio and updated the secio dep.

Stebalien commented on README.md:
```diff
@@ -9,25 +9,54 @@ go-libp2p-conn
+
+## Protocol overview
+
+The protocol is fairly straightforward: upon opening a connection, `go-multistream` is used to agree on plaintext (`"/plaintext/1.0.0"`) or encrypted (`"/secio/1.0.0"`). Both peers will only be open to use one or the other.
```
Plaintext will only be negotiated iff both peers have either `EncryptConnections` specified or haven't constructed their listeners/dialers with secret keys.

FiloSottile replied:
Tried to express that, rewrote this part.

Stebalien commented on README.md:
````diff
@@ -9,25 +9,54 @@ go-libp2p-conn
+### Tests
+
+```sh
+go test github.com/libp2p/go-libp2p-conn
````
Maybe add `make install` before this. Tests should be run after applying the gx transform.

Stebalien commented on dial.go:
```diff
@@ -64,15 +66,17 @@ func NewDialer(p peer.ID, pk ci.PrivKey, wrap WrapFunc) *Dialer {
-// Example: d.DialAddr(ctx, peer.Addresses()[0], peer)
-func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (iconn.Conn, error) {
+// Dial connects to a peer over a particular address.
+// raddr must be part of the remote peer Addresses().
```
I think this line is no longer relevant.

Stebalien commented on README.md:
````diff
@@ -9,25 +9,54 @@ go-libp2p-conn
 
 ```sh
-make install
+go get -u github.com/libp2p/go-libp2p-conn
````
Please leave this as `make install`. We use `gx` to manage our dependencies to prevent breaking changes in dependencies from breaking dependent packages.

FiloSottile replied:
There is no `make install` target in the Makefile.

Stebalien replied:
Ah. Good point. That should be `make deps`.

FiloSottile replied:
Done.

Stebalien reviewed (changes requested):
Forgot to review the documentation updates.

FiloSottile committed:
Address documentation review comments

FiloSottile commented:
Bump for merge?
